### PR TITLE
Fix static ghost behaviours

### PR DESCRIPTION
**Pull request by saurtron**
### Work done

- Add `leavesGhost` unitdef.
  - defaults to true for IsBuildingUnits
  - controls what units get unit attribute `leavesGhost` set to true on creation.
- Add `leavesGhost` unit attribute.
  - defaults to unitDef->leavesGhost
  - per unit toggle for static radar ghosts
  - lua `Get/SetUnitLeavesGhost` allows changing this at any time
  - controls:
    - radar wobbling when under radar
    - leaving static ghosts when out of radar or los (and having been in los and contradar before)
  - leaves a dead ghost behind when leavesGhost gets disabled and the unit had a static ghost, this way no information is leaked about unit whereabouts (only if 3rd parameter to SetUnitRadarGhost set to true).
    - this allows immobile units hopping on transports and looking like they're still there
- Add sample gadget `mobile_ghost_transport.lua`.

### Related issues

- https://github.com/beyond-all-reason/Beyond-All-Reason/issues/3617
- Fixes radar wobbling for units with leavesGhost true
- Fixes static radar ghosts being available just for buildings
- Fixes ghost behavior for transported units

### Remarks

- This approach allows control of which units leave static live/dead ghosts, and also disabling the ghosts behavior when the unit is put in a transport for example
- Dead ghost icons fixed at separate PR: https://github.com/beyond-all-reason/spring/pull/1887
- Could rename gameSetup->ghostedBuildings to staticGhosts (and export/accept also ghostedBuildings alias for now).
- Could rename live/deadGhostBuildings to live/deadStaticGhosts since it's not special for buildings now.
  - maybe change these at a different PR to not complicate this further

### Pending

- Fix ghosts not being generated when losing radar coverage of an area. (is this really happening, can't reproduce it now?)
  - not sure this is possible as part of this pr
  - might need some mechanism to facilitate drawing helpers (like defenseranges) for ghosts
    - atm this isn't handled properly at least in bar, not sure about zk or engine
      - ex: when an llt moves the range stays with the ghost, and then, if the llt new pos is spotted it won't draw a range

### Reference

```LUA
IsImmobileUnit -> (pathType == -1U && !canfly && speed <= 0.0f)
IsBuildingUnit -> (IsImmobileUnit() && !yardmap.empty())
IsGroundUnit   -> (pathType != -1U && !canfly)
```

### Screenshots

<image src="https://github.com/user-attachments/assets/87ac241b-357c-40c5-ae1d-1f357200d55d"></image>

### CODE REVIEW

**Comment by sprunk** [comment]
This just forces immobiles to have ghosts while leaving games no control. Add a new unit def tag, `bool leavesBuildingGhost`, and make ghosts look for that.

I dug a bit and I'm no longer sure whether that new tag should default to true for immobiles. See 30c07c56b85e05607a427875633dcb089ef61078, e573bdb6ab8121355f29d04b53fc030c2b2bd6c1, and https://springrts.com/mantis/view.php?id=6127.

**Comment by sprunk** [comment]
Would be nice to split this into two PRs: small one adding the boolean tag, and the big refactor + fixing minimap icons.

File: rts/Lua/LuaSyncedCtrl.cpp [hunk 1]
```diff
@@ -2684,6 +2686,35 @@ int LuaSyncedCtrl::SetUnitSeismicSignature(lua_State* L)
...
+
+	CUnit* unit = ParseUnit(L, __func__, 1);
+
+	if (unit == nullptr)
+		return 0;
+
+	bool prevValue = unit->leavesGhost;
+	unit->SetLeavesGhost(luaL_checkboolean(L, 2));
+	if (prevValue != unit->leavesGhost)
+		unitDrawer->UnitLeavesGhostChanged(unit, luaL_optboolean(L, 3, false));
```
Comment (sprunk) [inline rts/Lua/LuaSyncedCtrl.cpp:2714]: These two lines should probably be part of `unit->SetLeavesGhost` since otherwise unit drawer can fail to be notified of a change.

File: rts/Lua/LuaSyncedRead.cpp [hunk 2]
```diff
@@ -3850,6 +3851,22 @@ int LuaSyncedRead::GetUnitSeismicSignature(lua_State* L)
...
 
+/***
+ *
+ * @function Spring.GetUnitLeavesGhost
+ * @number unitID
+ * @treturn nil|number
+ */
+int LuaSyncedRead::GetUnitLeavesGhost(lua_State* L)
+{
+	const CUnit* const unit = ParseAllyUnit(L, __func__, 1);
```
Comment (sprunk) [inline rts/Lua/LuaSyncedRead.cpp:3862]: Should probably be allowed for enemy units too, since that is whose ghosts you're interested in.

File: rts/Rendering/Units/UnitDrawerData.cpp [hunk 3]
```diff
@@ -615,10 +616,25 @@ void CUnitDrawerData::RenderUnitDestroyed(const CUnit* unit)
 			// (the ref-counter saves us come deletion time)
 			savedData.deadGhostBuildings[allyTeam][gsoModel->type].push_back(gso);
 			gso->IncRef();
+			// remove prevlos for unit
+			if (u->losStatus[allyTeam] & LOS_PREVLOS)
+				u->losStatus[allyTeam] ^= LOS_PREVLOS;
```
Comment (sprunk) [inline rts/Rendering/Units/UnitDrawerData.cpp:621]: Why?

File: rts/Sim/Units/Unit.cpp [hunk 4]
```diff
@@ -555,7 +562,7 @@ float3 CUnit::GetErrorVector(int argAllyTeam) const
 	const int atSightMask = losStatus[argAllyTeam];
 
 	const int isVisible = 2 * ((atSightMask & LOS_INLOS  ) != 0 ||                  teamHandler.Ally(argAllyTeam, allyteam)); // in LOS or allied, no error
-	const int seenGhost = 4 * ((atSightMask & LOS_PREVLOS) != 0 && gameSetup->ghostedBuildings && unitDef->IsBuildingUnit()); // seen ghosted building, no error
+	const int seenGhost = 4 * ((atSightMask & LOS_PREVLOS) != 0 && leavesGhost); // seen ghosted immobiles, no error
```
Comment (sprunk) [inline rts/Sim/Units/Unit.cpp:565]: Ghosts are no longer tied to being an immobile.
```suggestion
	const int seenGhost = 4 * ((atSightMask & LOS_PREVLOS) != 0 && leavesGhost); // seen ghost, no error
```

File: rts/Lua/LuaSyncedCtrl.cpp [hunk 5]
```diff
@@ -2684,6 +2686,35 @@ int LuaSyncedCtrl::SetUnitSeismicSignature(lua_State* L)
...
+/***
+ * @function Spring.SetUnitLeavesGhost
+ *
+ * Change the unit leavesGhost attribute.
+ *
+ * Controls unit having static radar ghosts.
+ *
+ * @number unitID
+ * @bool leavesGhost
+ * @bool[opt] leaveDeadGhost leave a dead ghost behind if disabling and the unit had a live static ghost.
```
Comment (sprunk) [inline rts/Lua/LuaSyncedCtrl.cpp:2698]: It would be good if leaving a dead ghost was a standalone function. That way `SetUnitLeavesGhost` is atomic and you can leave dead ghosts around in other circumstances at will.

File: rts/Rendering/Units/UnitDrawerData.cpp (hunk 3 above)
Comment (saurtron) [reply rts/Rendering/Units/UnitDrawerData.cpp]: May not be very intuitive, but the prevlos status is "passed on" to the ghost. This is needed so the unit won't generate another ghost when/if leavesGhost is set to true again later on in the game.

File: rts/Sim/Units/Unit.cpp (hunk 4 above)
Comment (saurtron) [reply rts/Sim/Units/Unit.cpp:565]: Unsure about this, since this is just for ghosts of "generally immobile" units.

At least I don't see how this kind of ghosts could have a meaning in the game unless the unit has some kind of "generally immobile" assumption, even if it can sometimes be broken, like when a building is moved with some exception like being transported.

Mobile ones have a different kind of ghost.

Comment (sprunk) [reply rts/Sim/Units/Unit.cpp:565]: Can you point to the different kind of ghost for mobiles? Maybe handling for the two could be combined somehow.

Comment (saurtron) [reply rts/Sim/Units/Unit.cpp:565]: actually, looks like the mobile ones don't get drawn by engine itself, for bar it looks like it goes through unit_ghostradar_gl4.

Comment (saurtron) [reply rts/Sim/Units/Unit.cpp:565]: > actually, looks like the mobile ones don't get drawn by engine itself, for bar it looks like it goes through unit_ghostradar_gl4.

I suspect they could use the same path liveBuildingGhosts are using tho, they probably doing it for nothing

File: rts/Lua/LuaSyncedCtrl.cpp (hunk 1 above)
Comment (saurtron) [reply rts/Lua/LuaSyncedCtrl.cpp]: Ok, moved those.

File: rts/Lua/LuaSyncedRead.cpp (hunk 2 above)
Comment (saurtron) [reply rts/Lua/LuaSyncedRead.cpp:3869]: Not totally sure... synced probably gets it anyways, and also from unsynced you can get the game setup values by querying unitdef. Allowing unsynced GetUnitLeavesGhost for enemies may leak information and not convinced it's needed.

It could always be added if the need really arises, but pls say so if your are certain, probably not a big deal to expose it anyways.

File: rts/Lua/LuaSyncedCtrl.cpp (hunk 5 above)
Comment (saurtron) [reply rts/Lua/LuaSyncedCtrl.cpp:2859]: Regarding exposing as a standalone function, I would rather not do it for now, since I'm not sure it could lead to buggy corner cases. It'd be better to review as is and later add the independent lua method on top of this if wanted/needed.

File: rts/Lua/LuaSyncedRead.cpp (hunk 2 above)
Comment (sprunk) [reply rts/Lua/LuaSyncedRead.cpp:3869]: Sounds fine to keep as-is, it can always be exposed from synced as a public rules param.

File: rts/Lua/LuaSyncedCtrl.cpp (hunk 5 above)
Comment (sprunk) [reply rts/Lua/LuaSyncedCtrl.cpp:2859]: I thought about it some more and I think it's probably best to leave as-is for now.

**Review by sprunk — APPROVED** [review]
Haven't tested yet but looks acceptable.

**Comment by saurtron** [comment]
> Haven't tested yet but looks acceptable.

ok, in case you want to test I'm using this: https://github.com/saurtron/Beyond-All-Reason/commit/246a4c8387ba4b565e95cfd1213afc5c7843dd89

That one doesn't set the flag for nanos, but can be tested with llts. The idea is create an some llt and a transport with one team, try spotting some of the llts with the other team, then queue the first team transport to load the llt and unload far away, see what happens as observer team (should still see the ghost at the original place when the unit is carried away, instead of seeing the ghost moving with the transport to the new location)

**Comment by sprunk** [comment]
I tested and it works great.

**Comment by saurtron** [comment]
will merge soon, give a last review if you want

**Comment by saurtron** [comment]
> I don't remember what version I last approved and there were a bunch of changes since then, I'll re-review.

Ok, in case it helps, I think you tested d278ea1b629417a9dbd90a8b6fc5bffaccea2b6a "add some comments", I did add an example gadget, and also fixed a few corner cases since then.

**Comment by saurtron** [comment]
Hey @sprunk, did you manage to test this? added the following demo [nodrift.sdfz](https://github.com/saurtron/spring/raw/refs/heads/no-drift-for-immobiles-demo/testdata/nodrift.sdfz) in case it helps.

Uses bar master 2b42dfbb11 with the gadget included here and armnanotc set to leavesGhost = true.

**Review by lhog — APPROVED** [review]

**Review by sprunk — APPROVED** [review]
